### ISSUE TITLE
[3.0] Notification Response / answer

### ISSUE THREAD

**Issue by barryvdh** [thephpleague/omnipay#327]
Some gateways require a specific answer on the NotifyURL, eg. `OK` or sometimes something depending on the status/reference.

This would usually be the case for off-site payment, in the CompletePurchaseResponse.

I can't recall any that need a different status, other than 200, but we could have a `getNotificationResponse()` method which returns a PSR-7 response, or `getNotificationMessage`/`getNotificationReply`/`getNotificationAnswer` etc. Or just use the response `getNotificationResponse()->getBody()`. Along with an interface + `hasReply` so it can be checked for all gateways.

**Comment by kayladnls** [comment]
I'm a fan of the latter option. 

`getNotificationResponse()->getBody()` + `hasReply`

**Comment by judgej** [comment]
I've been getting my head around PSR-7 to see if I can create a gateway that is 100% PSR-7, and totally client agnostic. So far it makes requests and gets responses as PSR-7 messages and that works fine, with (very simple) adapters for Guzzle and Zend Diactoros (and some auto-discovery to pick up which is installed if you don't specify which you want to use).

Now, handling the Notifications I realised that Guzzle does not do this yet. There is no PSR-7 `ServerRequestInterface` support from Guzzle. Diactoros does it no problem. For this reason, if we want to support PSR-7 fully, we cannot rely on Guzzle being present. OmniPay at the moment does not care for PSR-7 server requests (requests coming into the server) - with little or no exception the gateways just need GET and POST data.

So the question is, are we extending to PSR-7 `ServerRequestInterface` support in OmniPay, or just using PSR-7 for `ServerRequest` and `ServerResponse`?

I'm still not convinced that Guzzle should be a requirement of OmniPay, especially since it is Guzzle version dependencies that are making upgrades to Laravel 5.2 difficult for many people, but I suspect the project is not ready for too big a leap into a more portable, but slightly more fiddly to use, package.

I'll carry on with my PSR-7 gateway to see if reality pans out as I hope, or is just an impractical pipe-dream :-)

**Comment by aimeos** [comment]
+1 for full PSR-7 support :-)

**Comment by barryvdh** [comment]
I'm not really sure if I understand the issue. We replaced the Request from Symfony with the ServerRequestInterface: https://github.com/thephpleague/omnipay/blob/59139cd0867a0f8aa1ed777b06de9daeeff7ff08/src/Common/Message/AbstractRequest.php#L111
Which is indeed done by Diactoros because Guzzle does not support it: https://github.com/thephpleague/omnipay/blob/59139cd0867a0f8aa1ed777b06de9daeeff7ff08/src/Common/AbstractGateway.php#L312

It's not really related to Guzzle imho (or Diactoros). Same as Guzzle, we rely on the interface, but use 1 implementation by default.

**Comment by judgej** [comment]
So long as we have this in composer,json, then both Guzzle and Diactoros are required. Whether using them is a "default" or not, there is a dependency created:

~~~json
"require": {
        "guzzlehttp/guzzle": "^6.0",
        "zendframework/zend-diactoros": "^1.1.0"
        ...
    },
~~~

Anyway, that is a side issue to this main issue. I *thought* we were locked to Guzzle 6, and that didn't support a `ServerRequestInterface` implementation. Now I realise to solve that we are locked into *two* HTTP libraries, and that is a step in further in the wrong direction IMO.

**Comment by barryvdh** [comment]
I understand the issue, but I just don't see a viable way to *not* require it, but still provide a user friendly setup. I know you and I can just be told to require a valid http driver and psr-7 implementation and inject it somewhere, but not sure if everyone thinks that's easy enough..

**Comment by judgej** [comment]
Yes, it is an extra composer require, assuming the package is not already installed. I'm trying something like this to discover what's installed and create a factory adapter as a result (this is still a bit experimental, and not OmniPay, but the hope is that I can wrap a thin OmniPay layer around it to create a gateway driver at a later date):

https://github.com/academe/SagePay-Integration/blob/PSR-7/src/Request/AbstractRequest.php#L107

**Comment by barryvdh** [comment]
Just to be clear, the gateways itself should never have knowledge about guzzle or diactoros. So even if we end up with conflicts soonish, it will only affect the common package, no need to update every single gateway.

**Comment by judgej** [comment]
Yes, totally.

**Comment by barryvdh** [comment]
Yeah but I mean, that the case already now with Omnipay v3.

**Comment by mkrauser** [comment]
> I understand the issue, but I just don't see a viable way to not require it, but still provide a user friendly setup. I know you and I can just be told to require a valid http driver and psr-7 implementation and inject it somewhere, but not sure if everyone thinks that's easy enough..

@barryvdh @judgej You are both right, including guzzle and diactoros as (hard) dependencies is not pretty. But if omnipay leaves it to the developer to install a client of his or her choosing, that would be more initial work and probably sometimes frustration if the dev does not read the documentation. 

Personally I would not like it, if omnipay "requires" diactoros and guzzle. But I don't have a really good solution either.

**Comment by aimeos** [comment]
There might be one possibility: Composer offers a "provide" section and meta packages. There's already a meta package for PSR-7: https://packagist.org/providers/psr/http-message. The only requirement would be that we convince the Guzzle and Zend teams to add this to their packages.

**Comment by mkrauser** [comment]
The right virtual package would be psr/http-message-implementation. Guzzle and Diactoros already provide that. 
But if that is required and no package which provides this, composer just complains that no package with this name or version is found, which is not very user friendly imho.

**Comment by judgej** [comment]
We probably also need to work on a PR for Guzzle to support a PSR-7 `ServerRequestInterface` implementation too. Without that we are needing Diactoros anyway, even if "choosing" (however we do that in the end) to use Guzzle.

**Comment by barryvdh** [comment]
So come back on the notification. there already is a NotificationInterface, but I don't think anyone uses that, right?

See http://packanalyst.com/class?q=Omnipay\Common\Message\NotificationInterface

 @sergej-koscejev @beinbm  Can you provide your actual use-case and how this might fit this issue?

**Comment by barryvdh** [comment]
@judgej Guzzle PSR-7 v1.3.0 now supports a ServerRequestInterface implementation https://github.com/guzzle/psr7/releases/tag/1.3.0 :)

**Comment by judgej** [comment]
Oh happy days. I just replaced this:

    $psr7ServerRequest = \Zend\Diactoros\ServerRequestFactory::fromGlobals();

with this:

    $psr7ServerRequest = \GuzzleHttp\Psr7\ServerRequest::fromGlobals();

in some test scripts, and the application did not even blink :-)

**Comment by judgej** [comment]
There actually are some slight variations between the `ServerRequest` messages of two implementations. Diactoros will return POST data in `getBody()` while GuzzleHttp returns an empt stream. Or maybe it is the casting of the body stream to a string that differs between the two. Either way, there are differences and I'm not sure which is correct at this stage.
